### DISCUSSION TITLE
SSL server options implementation

### DISCUSSION THREAD

**Discussion by sushantAtAws** [emqx/emqx#7486] [Q&A]
Hi, 
I am trying to understand the code base. I understand that the ssl listener conf in the emqx.conf will finally we translated to ssl conf (for instance https://www.erlang.org/doc/man/ssl.html#type-server_option). I am not able to wrap my head around how this conf is processed and where in code is it being applied as ssl options. 

For example, fail_if_no_peer_cert is one option that I can set in emqx.conf and has to be applied to ssl#server_option
Can someone help me understand how this plumbing is done in code? 

Thanks

**Comment by zmstone** [comment]
Hi @sushantAtAws 
Depends on the version.
For 4.x, the configs are parsed and mapped in the schema file.
e.g. https://github.com/emqx/emqx/blob/9a6ca6adf509e53cf9b8a433fc2d40b576bd219f/priv/emqx.schema#L2136-L2150

For 5.0, the schema is an Erlang module, `fail_if_no_peer_cert` is one of the config keys for listener.
https://github.com/emqx/emqx/blob/8f5812e61fba66c4c05939fd8cbc5d4056cf7ab2/apps/emqx/src/emqx_schema.erl#L2055
You'll have to search a bit in the code base to find where exactly the SSL options are used,
for example, for MQTT SSL listener, it should be emqx_listeners.erl module.
